doc/decisions/array.md

```diff
@@ -40,11 +38,30 @@ myarray/#5 = value5
   array = #5
 ```
 
+To lookup an array, first do `ksLookupByName (ks, "/myarray", 0)` on the parent.
+With the last index you get from its metadata `array`, iterate over the children.
+A cascading lookup on every individual child is also needed to make sure that overrides on individual
+elements are respected.
+
+For example:
+
+```
+spec:/myarray    # <- contains the specification for the array
+spec:/myarray/#  # <- contains the specification for the array elements
+
+dir:/myarray/#0  # <- not an array, just an override for user:/myarray/#
+
+user:/myarray    # <- with metadata array=#0, this would be the array we get
+user:/myarray/#0
+
+system:/myarray  # <- not found in cascading lookup, as user:/myarray exists
+```
+
```

Comment:
I noticed something I really don't like about this `array` concept: A `user:/` key can change the semantics of a (different) `system:/` key.

At the system level we have these keys:
```
system:/myarray              ; array=#0
system:/myarray/#0 = 8
system:/myarray/#1 = 12
```
That means for `root` and any other user `system:/myarray/#1` is not an array element.

But if some user adds this key:
```
user:/myarray                ; array=#1
```
Suddenly `system:/myarray/#1` is an array element, but only for this user. That goes against the idea that `system:/` keys are system-wide.

We don't have to enforce this (although `spec` might be able to), but we should at least formulate some rules against this.

Reply:
I clarified in accc55f7673b92b12f676ed042a187354c40955f

The first question is if this is legal to have `array=#0` if there are actually two array entries. Maybe this should already be rejected.

And in general: system is not the namespace to **enforce** system configuration but `spec` is. So yes, a user can put a non-array key somewhere which is an array in system.

Reply:
> if this is legal to have `array=#0` if there are actually two array entries.

By the current definition, the second entry (`system:/myarray/#1`) would _not_ be an array entry. Its just another key that looks like an array entry by coincidence. This exactly what I don't like about `#...` is only an array entry, if `array` on the parent says so.

> So yes, a user can put a non-array key somewhere which is an array in system.

First, the `spec` plugin would actually prevent that. Second, that is not the problem, here. Here a user turns a non-array key into an array key. So its exactly the other way around...